File: 2023/day09/solution.py
#!/usr/bin/env python3
import sys
from functools import cache

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case '-p1':
        PART1, PART2 = True, False
      case '-p2':
        PART1, PART2 = False, True
      case '-p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  histories = [[int(y) for y in x.split()] for x in f.read().splitlines()]


def part1():
  total = 0
  for history in histories:
    if DEBUG > 1: print(f'{history=}')
    #this feels horrible but idk yet how to optimise it
    # use history as the lowest level of deltas
    deltas: list[list[int]] = [history]
    # keep adding a layer until they're all zeroes
    #FIXME: the breakout condition is mine in case of bugs
    while any(x != 0 for x in deltas[-1]) or len(deltas) == 1:
      if DEBUG > 1: print(f'  append {deltas[-1]=}')
      deltas.append([y - x for x, y in zip(deltas[-1], deltas[-1][1:])])
    if DEBUG > 1: print(f'  {deltas=}')
    # care to extrapolate?
    for i in range(len(deltas) - 2, -1, -1):
      if DEBUG > 1: print(f'  {i=}')
      deltas[i].append(deltas[i][-1] + deltas[i + 1][-1])
    if DEBUG > 1: print(f'{deltas[0]=}')
    total += deltas[0][-1]
  print(f'part 1: {total}')


def part2():
  total = 0
  for history in histories:
    if DEBUG > 1: print(f'{history=}')
    # we can just reverse history instead of trying to pushleft on a list and having all kinds of perf issues
    deltas: list[list[int]] = [list(reversed(history))]
    while any(x != 0 for x in deltas[-1]) or len(deltas) == 1:
      if DEBUG > 1: print(f'  append {deltas[-1]=}')
      deltas.append([y - x for x, y in zip(deltas[-1], deltas[-1][1:])])
    if DEBUG > 1: print(f'  {deltas=}')
    for i in range(len(deltas) - 2, -1, -1):
      if DEBUG > 1: print(f'  {i=}')
      deltas[i].append(deltas[i][-1] + deltas[i + 1][-1])
    if DEBUG > 1: print(f'{deltas[0]=}')
    total += deltas[0][-1]
  print(f'part 1: {total}')


if PART1: part1()
if PART2: part2()

File: 2023/day09/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

const parseInput = () =>
  input
    .split('\n')
    .filter((line) => line.trim())
    .map((line) => line.split(/\s+/).map((token) => Number(token)));

const extrapolate = (row: number[]): number[] => {
  // this is definitely a clowncar approach but it's fast enough so w/e
  // solved something similar for some of the tile counts in 2023d21

  const deltas: number[][] = [row];
  // expand deltas until all same value
  while (new Set(deltas.slice(-1)[0]).size > 1) {
    deltas.push(
      deltas
        .slice(-1)[0]
        .map((value, i, row) => row[i + 1] - value)
        .slice(0, -1)
    );
  }
  // push extrapolated values
  let prevValue = 0;
  for (let i = deltas.length - 1; i > -1; --i) {
    const newValue = prevValue + deltas[i].slice(-1)[0];
    deltas[i].push(newValue);
    prevValue = newValue;
  }
  debug(1, { deltas });
  // return the first level (i.e row with an extrapolated value)
  return deltas[0];
};

const part1 = () => {
  const data = parseInput();
  const total = data.map((row) => extrapolate(row)).reduce((acc, row) => acc + row.slice(-1)[0], 0);
  console.log('part 1:', total);
};

const part2 = () => {
  const data = parseInput();
  const total = data.map((row) => extrapolate(row.toReversed())).reduce((acc, row) => acc + row.slice(-1)[0], 0);
  console.log('part 2:', total);
};

if (PART1) part1();
if (PART2) part2();
